[Tablegen] Add keyword `dump`.

fpetrogalli opened the pull request:
The keyword is intended for debugging purpose. It prints a message to stderr.

This patch is based on code originally written by Adam Nemet, and on the feedback received by the reviewers in
https://reviews.llvm.org/D157492.

fpetrogalli committed:
[Tablegen] Add keywork `dump`.

The keyword is intended for debugging purpose. It prints a message to
stderr.

This patch is based on code originally written by Adam Nemet, and on
the feedback received by the reviewers in
https://reviews.llvm.org/D157492.

fpetrogalli commented on llvm/lib/TableGen/TGParser.cpp:
```diff
@@ -4359,3 +4390,27 @@ LLVM_DUMP_METHOD void MultiClass::dump() const {
+  if (!consume(tgtok::semi))
+    return TokError("expected ';'");
+
+  SMLoc ConditionLoc = Lex.getLoc();
+  if (CurRec)
+    CurRec->addDump(ConditionLoc, Message);
+  else
+    addEntry(std::make_unique<Record::DumpInfo>(ConditionLoc, Message));
```
```suggestion
  SMLoc Loc = Lex.getLoc();
  if (CurRec)
    CurRec->addDump(Loc, Message);
  else
    addEntry(std::make_unique<Record::DumpInfo>(Loc, Message));
```

fpetrogalli committed:
Rename var. [NFC]

nhaehnle commented on llvm/docs/TableGen/ProgRef.rst:
```diff
@@ -1275,6 +1275,23 @@ be nested.
+.. productionlist::
+   Dump: "dump"  `string` ";"
+
+For example, it can be used in combination with `!repl` to investigate
```
Should be `!repr`, right?

fpetrogalli replied:
```suggestion
For example, it can be used in combination with `!repr` to investigate
```

nhaehnle commented on llvm/docs/TableGen/ProgRef.rst:
```diff
@@ -1275,6 +1275,23 @@ be nested.
+.. code-block:: text
+
+  multiclass MC<dag s> {
+    dump "s = " # !repl(s);
```
Same here: `!repr`

fpetrogalli replied:
```suggestion
    dump "s = " # !repr(s);
```

nhaehnle commented on llvm/lib/TableGen/Error.cpp:
```diff
@@ -170,4 +170,11 @@ void CheckAssert(SMLoc Loc, Init *Condition, Init *Message) {
+  if (auto *MessageInit = dyn_cast<StringInit>(Message))
+    PrintNote(MessageInit->getValue());
+  else
+    PrintNote("(dump message is not a string)");
```
Suggest using `!repr` as part of the message?

fpetrogalli replied:
```suggestion
    PrintNote("(dump message is not a string - use `!repr`)");
```

nhaehnle reviewed (commented):
Acked, modulo some minor nits and a suggestion.

Typo in the title: "keywork"

wangpc-pp commented on llvm/lib/TableGen/Error.cpp:
```diff
@@ -170,4 +170,11 @@ void CheckAssert(SMLoc Loc, Init *Condition, Init *Message) {
   }
 }
 
+void CheckDump(SMLoc Loc, Init *Message) {
```
This is not a `check` I think, it does the `dump`.
Change it to `Dump`, `DoDump` or other more precise name?

fpetrogalli replied:
Yep - I've gone with `dumpMessage`

wangpc-pp commented on llvm/lib/TableGen/Error.cpp:
```diff
@@ -170,4 +170,11 @@ void CheckAssert(SMLoc Loc, Init *Condition, Init *Message) {
 
+void CheckDump(SMLoc Loc, Init *Message) {
+  if (auto *MessageInit = dyn_cast<StringInit>(Message))
+    PrintNote(MessageInit->getValue());
```
```suggestion
    PrintNote(Loc, MessageInit->getValue());
```
Use `Loc` here?

wangpc-pp commented on llvm/lib/TableGen/Error.cpp:
```diff
@@ -170,4 +170,11 @@ void CheckAssert(SMLoc Loc, Init *Condition, Init *Message) {
+  if (auto *MessageInit = dyn_cast<StringInit>(Message))
+    PrintNote(MessageInit->getValue());
+  else
+    PrintNote("(dump message is not a string)");
```
And here.

fpetrogalli replied:
Done - but in this case I am printing a warning instead of note. I have added a test for this.

wangpc-pp commented on llvm/test/TableGen/dump.td:
```diff
@@ -0,0 +1,46 @@
+// CHECK-NEXT: dag X = (op op);
+// CHECK-NEXT: }
+defm X : MC<(op a)>;
+
```
Please add some tests for `dump`s in `foreach`, `class`, etc.

fpetrogalli replied:
Done!

wangpc-pp commented on llvm/lib/TableGen/TGParser.cpp:
```diff
@@ -4359,3 +4390,27 @@ LLVM_DUMP_METHOD void MultiClass::dump() const {
+  if (!consume(tgtok::semi))
+    return TokError("expected ';'");
+
+  SMLoc Loc = Lex.getLoc();
```
This `Loc` is not exact. Should it be the location of `dump` or message?

fpetrogalli replied:
Done - I have checked the lines in the tests.

wangpc-pp commented on llvm/lib/TableGen/TGParser.cpp:
```diff
@@ -498,6 +503,18 @@ bool TGParser::resolve(const std::vector<RecordsEntry> &Source,
         CheckAssert(E.Assertion->Loc, Condition, Message);
 
+    } else if (E.Dump) {
+      MapResolver R;
```
This code is duplicated with above.

fpetrogalli replied:
This code is needed to be able to dump the template parameters of a multiclass instance. In fact, if I remove it, tablegen crashes on the `dump.td` test at the function `llvm::TGParser::resolve`, when resolving the template argument `string s` of the multiclass `MC`.

fpetrogalli replied:
Hum - how could I have put the code for E.dump in the code for E.Assertion? They use different methods...

Anyway - if you are OK in keeping it as it is, please resolve the conversation :)

fpetrogalli committed:
Update llvm/docs/TableGen/ProgRef.rst

fpetrogalli committed:
Update llvm/docs/TableGen/ProgRef.rst

fpetrogalli committed:
Update llvm/lib/TableGen/Error.cpp

fpetrogalli committed:
s/CheckDump/dumpMessage

fpetrogalli committed:
Use the correct Loc.

fpetrogalli committed:
test for warning on not using a string.

fpetrogalli committed:
Print a warning instead of a note if the dump operand is not a string.

fpetrogalli committed:
Test if/foreach/defvar [NFCI].

fpetrogalli committed:
Check the use of dump in `class`, even through inhertance.

fpetrogalli committed:
Test dump on defset.

Added a comment to the place were we could make repr look recursively
into a list.

wangpc-pp commented on llvm/lib/TableGen/Record.cpp:
```diff
@@ -3119,6 +3133,16 @@ void Record::checkRecordAssertions() {
   }
 }
 
+void Record::checkRecordDumps() {
```
This name shoule be changed too.

wangpc-pp reviewed (commented):
LGTM but please wait for others.

wangpc-pp commented:
Oh! Don't forget to add a ReleaseNote!

fpetrogalli committed:
s/checkRecordDumps/emitRecordDumps/ [NFC]

fpetrogalli commented:
> Oh! Don't forget to add a ReleaseNote!

I'll prepare a separate PR for `dump` and `!repr`. Thanks

Artem-B commented on llvm/docs/TableGen/ProgRef.rst:
```diff
@@ -1275,6 +1275,23 @@ be nested.
+``dump`` --- print messages to stderr
+-------------------------------------
+
+A ``dump`` statement prints the input string to standard error
```
It may be useful to clarify when the printout actually happens. 
E.g. for `dump` inside a class, that would probably be during individual record instantiation.

This suggests that the printout order will not match that of the order of `dump()` in the source file. I wonder if we should print out some sort of location context -- e.g. the record or class name. Just a suggestion. I haven't tried it yet and don't know how usable it is in practice.

fpetrogalli replied:
Done, docs updated.

Artem-B commented on llvm/lib/TableGen/Error.cpp:
```diff
@@ -170,4 +170,12 @@ void CheckAssert(SMLoc Loc, Init *Condition, Init *Message) {
+  if (auto *MessageInit = dyn_cast<StringInit>(Message))
+    PrintNote(Loc, MessageInit->getValue());
+  else
+    PrintWarning(Loc, "dump message is not a string - use `!repr`");
```
Given that it's the only failure mode with the only actionable fix, should we magically wrap non-string inputs in `!repr` and allow `dump` to print out anything?

fpetrogalli replied:
Done

Artem-B commented on llvm/lib/TableGen/Record.cpp:
```diff
@@ -806,9 +806,12 @@ Init *UnOpInit::Fold(Record *CurRec, bool IsFinal) const {
+      } else {
+        // Otherwise, print the value of the variable.
+        //
+        // NOTE: we could look recursively !repr the elements of a
+        // list, but that could produce a lot of output when printing
+        // a defset.
```
I think "look" is out of place here.

fpetrogalli replied:
```suggestion
        // NOTE: we could recursively !repr the elements of a list, 
        // but that could produce a lot of output when printing a
        // defset.
```

Artem-B commented on llvm/test/TableGen/dump.td:
```diff
@@ -0,0 +1,86 @@
+  def A;
+}
+defm X : MC<(op a)>;
+
```
I'd add another `defm` . I wonder what the dump from multiple instances of in-multiclass dump would look like. Can we tell them apart somehow? I think they all will be attributed to the source location of the `dump` itself. I think we'll need some sort of reference to the instantiation point, too.

Not sure if it's worth it, but I think it would be useful. We can consider it as a follow-up change.

fpetrogalli replied:
Done (more on order of dump  and defs in a separate comment).

Artem-B reviewed (approved):
LGTM, with a few nits/suggestions.

fpetrogalli committed:
Update coment in  llvm/lib/TableGen/Record.cpp [NFC]

fpetrogalli committed:
Apply clang-format to latest commit. [NFC]

fpetrogalli committed:
Allow dumping of def and defvar directly.

fpetrogalli committed:
Remove unused variable. [NFC]

fpetrogalli committed:
Enforce dumpMessage to require a StringInit to print.

fpetrogalli committed:
[docs] Explain when the messages are instantiated. [NFC]

The text is inspired by the similar items in of `assert` keyword.

fpetrogalli committed:
Add test case with multiple nested `defm`. [NFC]

fpetrogalli commented:
@Artem-B 

>  I wonder what the dump from multiple instances of in-multiclass dump would look like. Can we tell them apart somehow? I think they all will be attributed to the source location of the dump itself. I think we'll need some sort of reference to the instantiation point, too.

I like this a lot - it can get nasty if someone start using dump statements across layers of class/multiclass indirection.
I actually had the same issue with assert, which does not come as surprise given the similarities with dump.

> Not sure if it's worth it, but I think it would be useful. We can consider it as a follow-up change.

If this is OK with you, I will create an issue and link it to this PR.

fpetrogalli committed:
Fix code formmatting [NFC]

fpetrogalli commented:
@Artem-B - gentle ping,  I'd like to get a final blessing from you.

Francesco

Artem-B commented on llvm/docs/TableGen/ProgRef.rst:
```diff
@@ -1275,6 +1275,37 @@ be nested.
+
+* At top level, the message is printed immediately.
+
+* In a record definition, the message is saved and all messages are
+  printed after the record is completely built.
+
+* In a class definition, the messages are saved and inherited by all
+  the subclasses and records that inherit from the class. The messages
+  are then printed when the records are completely built.
+
+* In a multiclass definition, the messages are saved with the other
+  components of the multiclass and then printed each time the
+  multiclass is instantiated with ``defm``.
```
`message is saved` may not be the best way to describe things here.
When we use `dump()` in a class/multiclass, we do not have the "message" at that point, as we have no instances yet. It's when we instantiate the record, the dump and its argument get evaluated and, presumably, printed.

You may want to rephrase the doc accordingly.

It may actually collapse to something like "dump() within record/class/multiclass gets evaluated at each instantiation point of the containing record" which would cover all three cases here.

fpetrogalli replied:
Done, thank you!

Artem-B reviewed (approved):
LGTM, with one suggestion for the docs.

fpetrogalli committed:
Update docs after review feedback in https://github.com/llvm/llvm-project/pull/68793#discussion_r1364289681
